### ISSUE TITLE
[py2py3] Python3 Rucio client fails with NameError: 'imp'

### ISSUE THREAD

**Issue by todor-ivanov** [dmwm/WMCore#10525]
**Impact of the bug**
Python3 env and anything depending on Rucio client

**Describe the bug**
This issue is a follow up of this one: https://github.com/dmwm/WMCore/issues/10513. Once we have fixed the lack of `exceptions` module in python3 we have faced another issue related to the python3 version of the Rucio Client [1]

[1]
```
[dmwm@19ed0d6b7f55 WMCore]$ nosetests test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/LogArch_t.py
EEEEEEEE
======================================================================
ERROR: testExecutorDoesntDetonate (WMCore_t.WMSpec_t.Steps_t.Executors_t.LogArch_t.LogArchiveTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/dmwm/wmcore_unittest/WMCore/test/python/WMCore_t/WMSpec_t/Steps_t/Executors_t/LogArch_t.py", line 119, in setUp
    taskMaker.processWorkload()
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMSpec/Makers/TaskMaker.py", line 127, in processWorkload
    sandboxCreator.makeSandbox(self.workdir, self.workload)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMRuntime/SandboxCreator.py", line 108, in makeSandbox
    fetcherInstances = list(map(getFetcher, fetcherNames))
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMSpec/Steps/StepFactory.py", line 196, in getFetcher
    return _FetcherFactory.loadObject(fetcherName)
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMFactory.py", line 58, in loadObject
    module = __import__(module, globals(), locals(), [classname])
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/WMSpec/Steps/Fetchers/PileupFetcher.py", line 18, in <module>
    from WMCore.Services.Rucio.Rucio import Rucio
  File "/home/dmwm/wmcore_unittest/WMCore/src/python/WMCore/Services/Rucio/Rucio.py", line 17, in <module>
    from rucio.client import Client
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/__init__.py", line 20, in <module>
    from .client import Client  # noqa: F401
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/client.py", line 30, in <module>
    from rucio.client.accountclient import AccountClient
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/accountclient.py", line 39, in <module>
    from rucio.client.baseclient import BaseClient
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/baseclient.py", line 66, in <module>
    from rucio.common.utils import build_url, get_tmp_dir, my_key_generator, parse_response, ssh_sign, setup_logger
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/common/utils.py", line 107, in <module>
    imp.find_module(extra_module)
NameError: name 'imp' is not defined
```

**How to reproduce it**
Run the unit test in question unther the python3 environment inside the python3 unit tests docker container for WMCore.

**Expected behavior**
To have that unit test running with no Errors.

**Additional context and error message**
Add any other context about the problem here, like error message and/or traceback. You might want to use triple back ticks to properly format it.

**Comment by todor-ivanov** [comment]
It seems this time the error is not buried inside the WMCore code [1], but more related to either the environment or the Rucio py3 client itself.

FYI @amaltaro 

[1]
```
[dmwm@15167945c033 WMCore]$ rucio 
Traceback (most recent call last):
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/bin/rucio", line 65, in <module>
    from rucio.client import Client
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/__init__.py", line 20, in <module>
    from .client import Client  # noqa: F401
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/client.py", line 30, in <module>
    from rucio.client.accountclient import AccountClient
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/accountclient.py", line 39, in <module>
    from rucio.client.baseclient import BaseClient
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/client/baseclient.py", line 66, in <module>
    from rucio.common.utils import build_url, get_tmp_dir, my_key_generator, parse_response, ssh_sign, setup_logger
  File "/home/dmwm/unittestdeploy/wmagent/1.4.9.pre3/sw/slc7_amd64_gcc630/external/py3-rucio-clients/1.24.2.post1-comp/lib/python3.8/site-packages/rucio/common/utils.py", line 107, in <module>
    imp.find_module(extra_module)
NameError: name 'imp' is not defined

```

**Comment by klannon** [comment]
What's strange is that according to Google, `imp` is a standard Python library. It's been deprecated since 3.4, but isn't supposed to be removed until 3.12 as I understand it.

**Comment by mapellidario** [comment]
#### TL;DR

`rucio` version [v1.25.5](https://github.com/rucio/rucio/releases/tag/1.25.5) seems to fix our problem

#### extended version

In order to close #9974, I had to have a look at `imp` and `importlib`. In `WMCore/Configuration.py` we are successfully using `imp` in both py2 and py3 env. 

So this feels like a problem with `rucio/common/utils.py` itself. I had a peek at their code, at the beginning of the `client.py` they select the importing library with

```python
try:
    import importlib
    importlib.util.find_spec('')
except AttributeError:
    import imp
```

Since in python3 we have a fully working `importlib`, it does import `imp`. Then later, they do

```python
for extra_module in EXTRA_MODULES:
    if 'imp' in sys.modules:
        try:
            imp.find_module(extra_module)
            EXTRA_MODULES[extra_module] = True
        except ImportError:
            EXTRA_MODULES[extra_module] = False

```

since we provide imp with our py3 env, they try to use `imp` without having it imported before.

So, if we apply this change

```diff
- try:
-    import importlib
-    importlib.util.find_spec('')
-except AttributeError:
-    import imp
+import imp
```

to both `rucio/common/utils.py` and `rucio/client/baseclient.py`, then we can successfully import the rucio client

```plaintext
[dmwm@b36d67c2cbbc ~]$ python3
Python 3.8.2 (default, Jan 22 2021, 17:57:37) 
[GCC 6.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from rucio.client import Client
>>> 
```

We are lucky: it feels like that this PR already solved our problem https://github.com/rucio/rucio/pull/4373/files

**Comment by todor-ivanov** [comment]
Hi @amaltaro 
Here is the cmsdist PR that is supposed to solve the issue:
https://github.com/cms-sw/cmsdist/pull/6959

How about the docker image for the unit tests? Should we wait until it is rebuild and tested before we close this issue?

**Comment by amaltaro** [comment]
Thanks for providing this PR, Todor. I think it's best to test it first though, it isn't clear to me whether that actually fixes the problem (and that catch on `AttributeError` doesn't look right to me). I will get it tested today and update this PR.

**Comment by todor-ivanov** [comment]
Thank @amaltaro ! Totally agree with you.

**Comment by amaltaro** [comment]
It turns out I cannot reproduce this error with a `wmagentpy3` deployment, only in the docker image (which also contains wmcorepy3-devtools). I suggest we test it in the docker image once that rucio-clients spec has been updated, otherwise it would be way too much work now.

Nonetheless, I tried to update that package in the docker image and - unless I made some mistake - the rucio-clients==1.25.5 indeed seems to resolve that problem. I created this basic documentation for pip installation within the python3 container:
https://github.com/dmwm/WMCore/wiki/Docker-and-WMCore#install-python3-packages-via-pip

@todor-ivanov I'm going to merge your cmsdist PR, then later today we create new RPMs and docker image and we can run a final test to make sure things are working properly. So I suggest we keep it opened for a little longer.

Meanwhile, can you please create another PR updating our list of required python3 packages
https://github.com/dmwm/WMCore/blob/master/requirements_py3.txt#L15
?

**Comment by amaltaro** [comment]
And the rucio-clients update fixed this issue. The `LogArch_t` above still fails, but with a different error now. However, things work fine when testing the Rucio_t module itself:
```
[dmwm@c208bd73a58f WMCore]$ nosetests test/python/WMCore_t/Services_t/Rucio_t/Rucio_t.py:RucioTest.testWhoAmI
.
----------------------------------------------------------------------
Ran 1 test in 1.678s

OK
```

This issue can now be closed via:
https://github.com/dmwm/WMCore/pull/10567
and
cms-sw/cmsdist#6959

Thank you very much, Todor.
